File: vm/zway.go
package vm

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/dop251/goja"
	"github.com/orktes/huessimo/zway"
)

var zWayMap = map[string]*zway.ZWay{}

func (vm *VM) initZWay() {

	for _, zWay := range zWayMap {
		zWay.Stop()
	}

	vm.Set("_init_zway", func(call goja.FunctionCall) goja.Value {
		id := call.Argument(0).String()
		zwayConfigStr := call.Argument(1).String()

		config := &zway.Config{}
		json.Unmarshal([]byte(zwayConfigStr), config)

		zWay, err := zway.New(*config)
		if err != nil {
			panic(err)
		}

		zWayMap[id] = zWay

		zWay.Devices.AddNewDeviceListener(func(device *zway.Device) {
			deviceStr, err := json.Marshal(device)
			if err != nil {
				log.Println(err)
				return
			}

			_, err = vm.RunString(fmt.Sprintf(`require('zway')._device_added("%s", %s);`, id, deviceStr))
			if err != nil {
				log.Println(err)
			}

			device.Metrics.Level.AddValueChangeListener(func(level *zway.Level) {
				valueStr, err := json.Marshal(level)
				if err != nil {
					log.Println(err)
					return
				}
				_, err = vm.RunString(fmt.Sprintf(`require('zway')._value_change("%s", "%s", %s);`, id, device.ID, valueStr))
				if err != nil {
					log.Println(err)
				}
			})
		})

		return goja.Null()
	})

}

File: vm/zway.js
var EventEmitter = require('eventemitter');
var _ = require('lodash');

var zWayId = 0;
var gateways = {};

var ZWayDevice = function (data) {
  _.extend(this, data);
};

ZWayDevice.prototype = _.extend({}, EventEmitter.prototype, {
  _setLevel: function (value) {
    this.metrics.level = value;
    this.emit('level_change', value);
  }
});

var ZWay = function (config) {
  this.id = zWayId++;
  this.devices = {};

  gateways[this.id] = this;

  _init_zway(this.id, JSON.stringify({
    Hostname: config.hostname,
    Port: config.port,
    PollTimeout: (config.poll_timeout || 2) * 1000000000, // Convert seconds to nanoseconds
    Username: config.username,
    Password: config.password
  }));
};

ZWay.prototype =  _.extend(EventEmitter.prototype, {
  _addDevice: function (device) {
    device = new ZWayDevice(device)
    this.devices[device.id] = device;
    this.emit('device_added', device);
  },
});

module.exports.ZWay = ZWay;
module.exports._device_added = function (id, device) {
  gateways[id]._addDevice(device);
};

module.exports._value_change = function (id, deviceId, value) {
  gateways[id].devices[deviceId]._setLevel(value);
};
